autotools build system not properly tracking dependencies of mod files

This was pointed out by @skosukhin in #238.

As demonstrated by the following, mod file dependencies are not being properly tracked.

Sergey has also given us the solution, I will put a PR up with it shortly.

```
(base) ed@mikado:~/netcdf-fortran/fortran$ ls *.mod
netcdf4_f03.mod  netcdf4_nc_interfaces.mod  netcdf4_nf_interfaces.mod  netcdf_f03.mod  netcdf_fortv2_c_interfaces.mod  netcdf.mod  netcdf_nc_data.mod  netcdf_nc_interfaces.mod  netcdf_nf_data.mod  netcdf_nf_interfaces.mod  typesizes.mod
(base) ed@mikado:~/netcdf-fortran/fortran$ rm netcdf4_f03.mod 
(base) ed@mikado:~/netcdf-fortran/fortran$ make
make  all-am
make[1]: Entering directory '/home/ed/netcdf-fortran/fortran'
make[1]: Nothing to be done for 'all-am'.
make[1]: Leaving directory '/home/ed/netcdf-fortran/fortran'
(base) ed@mikado:~/netcdf-fortran/fortran$ make netcdf4_f03.mod 
make: Nothing to be done for 'netcdf4_f03.mod'.
(base) ed@mikado:~/netcdf-fortran/fortran$ ls *.mod
netcdf4_nc_interfaces.mod  netcdf4_nf_interfaces.mod  netcdf_f03.mod  netcdf_fortv2_c_interfaces.mod  netcdf.mod  netcdf_nc_data.mod  netcdf_nc_interfaces.mod  netcdf_nf_data.mod  netcdf_nf_interfaces.mod  typesizes.mod

```

Similarly for tests.mod in the nf_tests directory:

```
ed@mikado:~/netcdf-fortran/nf_test$ make check
make  check-am
make[1]: Entering directory '/home/ed/netcdf-fortran/nf_test'
make  nf_test tst_int64 tst_f77_v2 ftest nf03_test f03test tst03_f77_v2
make[2]: Entering directory '/home/ed/netcdf-fortran/nf_test'
make[2]: 'nf_test' is up to date.
make[2]: 'tst_int64' is up to date.
make[2]: 'tst_f77_v2' is up to date.
make[2]: 'ftest' is up to date.
gfortran -DPACKAGE_NAME=\"netCDF-Fortran\" -DPACKAGE_TARNAME=\"netcdf-fortran\" -DPACKAGE_VERSION=\"4.5.3-development\" -DPACKAGE_STRING=\"netCDF-Fortran\ 4.5.3-development\" -DPACKAGE_BUGREPORT=\"support-netcdf@unidata.ucar.edu\" -DPACKAGE_URL=\"\" -DPACKAGE=\"netcdf-fortran\" -DVERSION=\"4.5.3-development\" -DNF_RELAX_COORD_BOUND=1 -DENABLE_CDF5=1 -DHAVE_F2008=1 -DHAVE_TS29113_SUPPORT=1 -DTEMP_LARGE=\".\" -DSTDC_HEADERS=1 -DHAVE_SYS_TYPES_H=1 -DHAVE_SYS_STAT_H=1 -DHAVE_STDLIB_H=1 -DHAVE_STRING_H=1 -DHAVE_MEMORY_H=1 -DHAVE_STRINGS_H=1 -DHAVE_INTTYPES_H=1 -DHAVE_STDINT_H=1 -DHAVE_UNISTD_H=1 -DHAVE_DLFCN_H=1 -DLT_OBJDIR=\".libs/\" -DNF_INT1_T=byte -DNF_INT2_T=integer\*2 -DNF_INT8_T=integer\*8 -DNF_INT1_IS_C_SIGNED_CHAR=1 -DNF_INT2_IS_C_SHORT=1 -DNF_INT8_IS_C_LONG_LONG=1 -DNF_INT_IS_C_INT=1 -DNF_REAL_IS_C_FLOAT=1 -DNF_DOUBLEPRECISION_IS_C_DOUBLE=1 -DNCBYTE_T=byte -DNCSHORT_T=integer\*2 -DHAVE_LIBM=1 -DHAVE_NETCDF_H=1 -DHAVE_NC_DEF_OPAQUE=1 -DHAVE_NCCREATE=1 -DHAVE_NC_SET_LOG_LEVEL=1 -DHAVE_OC_OPEN=1 -DHAVE_NC_GET_CHUNK_CACHE_INTS=1 -DUSE_NETCDF4=1 -DLOGGING=1 -DHAVE_SYS_PARAM_H=1 -DHAVE_LOCALE_H=1 -DHAVE_STDIO_H=1 -DHAVE_STDARG_H=1 -DHAVE_ERRNO_H=1 -DHAVE_CTYPE_H=1 -DHAVE_FCNTL_H=1 -DHAVE_MALLOC_H=1 -DHAVE_STDLIB_H=1 -DHAVE_STRING_H=1 -DHAVE_STRINGS_H=1 -DHAVE_UNISTD_H=1 -DHAVE_SYS_STAT_H=1 -DHAVE_GETOPT_H=1 -DHAVE_SYS_TIME_H=1 -DHAVE_SYS_RESOURCE_H=1 -DHAVE_DIRENT_H=1 -DSTDC_HEADERS=1 -DHAVE_SYS_WAIT_H=1 -DTIME_WITH_SYS_TIME=1 -DHAVE_VPRINTF=1 -DHAVE__BOOL=1 -DHAVE_STDBOOL_H=1 -DHAVE_ALLOCA_H=1 -DHAVE_ALLOCA=1 -DHAVE_SSIZE_T=1 -DHAVE_PTRDIFF_T=1 -DSIZEOF_SHORT=2 -DSIZEOF_INT=4 -DSIZEOF_LONG=8 -DSIZEOF_FLOAT=4 -DSIZEOF_DOUBLE=8 -DSIZEOF_OFF_T=8 -DSIZEOF_SIZE_T=8 -DSIZEOF_PTRDIFF_T=8 -I.   -I/usr/local/netcdf-c-4.7.4_hdf5-1.10.6/include -I../fortran -g -Wall -c -o test03_get.o test03_get.F
test03_get.F:10:11:

         USE tests
           1
Fatal Error: Can't open module file ‘tests.mod’ for reading at (1): No such file or directory
compilation terminated.
Makefile:828: recipe for target 'test03_get.o' failed
make[2]: *** [test03_get.o] Error 1
make[2]: Leaving directory '/home/ed/netcdf-fortran/nf_test'
Makefile:1179: recipe for target 'check-am' failed
make[1]: *** [check-am] Error 2
make[1]: Leaving directory '/home/ed/netcdf-fortran/nf_test'
Makefile:1182: recipe for target 'check' failed
make: *** [check] Error 2
(base) ed@mikado:~/netcdf-fortran/nf_test$ make tests.mod
make: Nothing to be done for 'tests.mod'.
(base) ed@mikado:~/netcdf-fortran/nf_test$ 
```

With the fixes that Sergey has provided, both of these situations are correctly handled: if a mod file is deleted, the build system is aware of it and automatically rebuids it as needed.

PR up shortly.

--- a/fortran/Makefile.am
+++ b/fortran/Makefile.am
@@ -92,8 +92,8 @@ module_netcdf_f03.lo: netcdf_nf_data.mod netcdf_nf_interfaces.mod
 nf_nc.lo: netcdf_nc_interfaces.mod
 
 # Mod files are built and then installed as headers.
-MODFILES = typesizes.mod netcdf_nc_data.mod netcdf_nf_data.mod	\
-netcdf_nc_interfaces.mod netcdf_nf_interfaces.mod
+MODFILES = typesizes.mod netcdf_nc_data.mod netcdf_nf_data.mod		\
+netcdf_nc_interfaces.mod netcdf_nf_interfaces.mod netcdf_f03.mod
 
 # Add our convenience libraries to the netCDF fortran library.
 libnetcdff_la_LIBADD = libtypeSizes.la libnetcdf_nc_data.la	\
@@ -156,6 +156,20 @@ BUILT_SOURCES = $(MODFILES)
 
 # Mod files are installed, but not distrubuted.
 nodist_fmod_DATA = $(MODFILES)
+
+# If a module file is accidentally deleted, delete the corresponding
+# object and rebuild it.
+$(MODFILES):
+	@if test -z $<; then \
+	  echo "Cannot find Fortran source file declaring module '$(basename $(@F))'." >&2; \
+	elif test ! -f $@; then \
+	  rm -f $<; $(MAKE) $(AM_MAKEFLAGS) $<; \
+	fi
+
+# If the rule above rebuilds an object in a subprocess of make, the
+# current process might not know about it. Therefore, to enforce
+# regeneration of the library, we introduce the following dependency.
+EXTRA_libnetcdff_la_DEPENDENCIES = $(MODFILES)
 
 # The netcdf.inc file is built on the user machine, and installed as
 # a header.

--- a/fortran/Makefile.am
+++ b/fortran/Makefile.am
@@ -200,5 +200,5 @@ endif
 EXTRA_DIST = $(COMMON_CODES) $(NETCDF4_CODES) $(NETCDF3_CODES) CMakeLists.txt	\
 netcdf2.inc netcdf3.inc netcdf4.inc
 
-CLEANFILES = *.mod netcdf.inc
+MOSTLYCLEANFILES = *.mod netcdf.inc
 

--- a/nf_test/Makefile.am
+++ b/nf_test/Makefile.am
@@ -45,8 +45,9 @@ noinst_LTLIBRARIES = libtests.la
 # Each convenience library depends on its source.
 libtests_la_SOURCES = module_tests.F90
 
-# Each mod file depends on the .o file.
+# Rebuild the tests.mod file if needed.
 tests.mod: module_tests.$(OBJEXT)
+	@if test ! -f $@; then rm -f $<; $(MAKE) $<; fi
 
 # Mod files are built sources.
 BUILT_SOURCES = tests.mod

--- a/nf_test/Makefile.am
+++ b/nf_test/Makefile.am
@@ -143,11 +143,16 @@ endif #USE_SED
 EXTRA_DIST = test_get.F test_put.F fills.cdl ref_fills.nc	\
 CMakeLists.txt
 
-# Cleaning up files created during the process.
-CLEANFILES = scratch*.nc test*.nc copy*.nc tst_*.nc ftst_*.nc	\
-f90tst_*.nc f03test*.nc *.tmp *.tmp2 fort.* tst03*.nc	\
-f03tst*.nc ftest.nc fcopy.nc *.mod $(F03_TEST_CODES)
+# As per autotools convention, mod files get mostly cleaned.
+MOSTLYCLEANFILES = *.mod  
 
+# Cleaning up files created during the process.
+CLEANFILES = scratch*.nc test*.nc copy*.nc tst_*.nc ftst_*.nc		\
+f90tst_*.nc f03test*.nc *.tmp *.tmp2 fort.* tst03*.nc f03tst*.nc	\
+ftest.nc fcopy.nc $(F03_TEST_CODES)
+
+# This file gets copied to build dir by configure.
 DISTCLEANFILES = fills.nc
 
+# These are shipped with code, but can be rebuilt from m4 if needed.
 MAINTAINERCLEANFILES = test_get.F test_put.F

